include/deal.II/base/patterns_tools.h

```diff
@@ -0,0 +1,682 @@
+     *
+     * A class with list_rank::value = 0 is either elementary or a
+     * map. A class with map_rank::value = 0 is either a List compatible
+     * class, or an elementary type.
+     *
+     * Elementary types are not compatible with Patterns::List, but
+     * non elementary types, like Point<dim>(), or std::complex<double>, are
+     * compatible with the List type. Adding more compatible types is a matter
+     * of adding a specialization of this struct for the given type.
+     *
+     * @author Luca Heltai, 2017
+     */
+    template<class T, class Enable=void>
+    struct RankInfo
+    {
+      static constexpr int list_rank = 0;
+      static constexpr int map_rank = 0;
+    };
+  }
+
+  // Arithmetic types
+  template<class T>
+  struct Convert<T, typename std::enable_if<std::is_arithmetic<T>::value>::type>
+  {
+
+    static std::unique_ptr<PatternBase> to_pattern()
+    {
+      if (std::is_integral<T>::value)
+        return std_cxx14::make_unique<Integer>(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+      else if (std::is_floating_point<T>::value)
+        return std_cxx14::make_unique<Double>(-std::numeric_limits<T>::max(), std::numeric_limits<T>::max());
+      else
+        {
+          AssertThrow(false, ExcNotImplemented());
+          return std::unique_ptr<PatternBase>();
+        }
+    }
+
+    static std::string to_string(const T &value, const std::unique_ptr<PatternBase> &p = Convert<T>::to_pattern())
+    {
+      std::string str;
+      if (std::is_same<T, unsigned char>())
+        str = std::to_string((int)value);
+      else
+        str = std::to_string(value);
+      AssertThrow(p->match(str), ExcNoMatch(str, *p));
+      return str;
+    }
+
+    static T to_value(const std::string &s,
+                      const std::unique_ptr<PatternBase> &p = Convert<T>::to_pattern())
+    {
+      AssertThrow(p->match(s), ExcNoMatch(s, *p));
+      std::istringstream is(s);
+      T value;
+      if (std::is_same<T,unsigned char>::value)
+        {
+          int i;
+          is >> i;
+          value=i;
+        }
+      else
+        is >> value;
+      AssertThrow(!is.fail(), ExcMessage("Failed to convert from \"" + s + "\" to the type \""
+                                         +boost::core::demangle(typeid(T).name()) + "\""));
+      return value;
+    }
+  };
+
+
+  namespace internal
+  {
+    const std::array<std::string, 5> default_list_separator {{" ",  ","  ,  ";"  ,  "|"  ,   "%"}};
+    const std::array<std::string, 5> default_map_separator {{" ",  ":"  ,  "="  ,  "@"  ,   "#"}};
+
+    //specialize a type for all of the STL containers and maps
+    template <typename T>       struct is_stl_container:std::false_type {};
+    template <typename T, std::size_t N> struct is_stl_container<std::array    <T,N>>    :std::true_type {};
+    template <typename... Args> struct is_stl_container<std::vector            <Args...>>:std::true_type {};
```

Comment:
maybe put a space around the `:` that denotes derivation here, to ease readability? same below, of course.